actors/hierarchical_sca/src/atomic/mod.rs

```diff
@@ -0,0 +1,150 @@
+use cid::multihash::Code::Blake2b256;
+use cid::multihash::MultihashDigest;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use fvm_ipld_encoding::{serde_bytes, tuple::*, Cbor, RawBytes, DAG_CBOR};
+use fvm_shared::MethodNum;
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+
+use crate::tcid::{TCid, THamt};
+
+/// MethodNum to lock some state in an actor
+/// This methods are only supported in actors
+/// that support atomic executions.
+pub const METHOD_LOCK: MethodNum = 2;
+/// MethodNum used to trigger the merge of an input with
+/// other input locked states.
+pub const METHOD_MERGE: MethodNum = 3;
+/// MethodNum called to signal the abortion of an atomic execution
+/// and the unlock of all locked states in the actor for the execution
+pub const METHOD_ABORT: MethodNum = 4;
+/// MethodNum to trigger the merge of the output of an execution
+/// into the state of an actor, and the unlock of all locked states.
+pub const METHOD_UNLOCK: MethodNum = 5;
+
+/// Trait that determines the functions that need to be implemented by
+/// a state object to be lockable and be used in an atomic execution.
+///
+/// Different strategies may be used to merge different locked state to
+/// prepare the actor state for the execution, and for the merging of the
+/// output of the execution to the original state of the actor.
+pub trait LockableState<S: Serialize + DeserializeOwned> {
+    /// Merge a locked state (not necessarily the output) to the current state.
+    fn merge(other: Self) -> anyhow::Result<()>;
+    /// Merge the output of an execution to the current state.
+    fn merge_output(other: Self) -> anyhow::Result<()>;
+}
+
+/// Trait that specifies the interface of an actor state able to support
+/// atomic executions.
+pub trait LockableActorState<T>
+where
+    T: Serialize + DeserializeOwned + LockableState<T>,
+{
+    /// Map with all the locked state in the actor uniquely identified through
+    /// their Cid.
+    fn locked_map_cid() -> TCid<THamt<Cid, LockedState<T>>>;
+    /// Returns the output state of an execution from the current state
+    /// of the actor according to the input parameters.
+    fn output(params: LockParams) -> LockedState<T>;
```

Comment:
```suggestion
    /// Map with all the locked state in the actor uniquely identified through
    /// their Cid.
    fn locked_map_cid(&self) -> TCid<THamt<Cid, LockedState<T>>>;
    /// Returns the output state of an execution from the current state
    /// of the actor according to the input parameters.
    fn output(&self, params: LockParams) -> LockedState<T>;
```
At least I'd think it's not some kind of type level function.

Reply:
Sorry, I might just be biased by my earlier experience with Actors (mainly the Akka project in Scala) being individuals. I seem to recall that actors in Filecoin were unit structs. I'll see how this works out further down.

Reply:
No, you are right. Again, this is just me copy-pasting from the Go interface and leave it as a placeholder for now. Thanks :pray: